File: leetcode-cn/Python/36.有效的数独.py
#
# @lc app=leetcode.cn id=36 lang=python3
#
# [36] 有效的数独
#

# @lc code=start
from typing import List


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        hang_list = [[0] * 9 for _ in range(9)]
        lie_list = [[0] * 9 for _ in range(9)]
        gong_list = [[0] * 9 for _ in range(9)]

        for i in range(9):
            for j in range(9):
                if board[i][j] == '.':
                    continue

                num = int(board[i][j]) - 1

                if hang_list[i][num] == 1:
                    return False
                else:
                    hang_list[i][num] = 1

                if lie_list[j][num] == 1:
                    return False
                else:
                    lie_list[j][num] = 1

                id = (i // 3) * 3 + (j // 3)
                if gong_list[id][num] == 1:
                    return False
                else:
                    gong_list[id][num] = 1

        return True


# @lc code=end

File: leetcode-cn/C++/36.有效的数独.cpp
/*
 * @lc app=leetcode.cn id=36 lang=cpp
 *
 * [36] 有效的数独
 */

// @lc code=start
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board)
    {
        vector<vector<bool>> hang(9, vector<bool>(9, false));
        vector<vector<bool>> lie(9, vector<bool>(9, false));
        vector<vector<bool>> jiu(9, vector<bool>(9, false));

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') {
                    continue;
                }
                int num = board[i][j] - '1';

                if (hang[i][num]) {
                    return false;
                }

                if (lie[j][num]) {
                    return false;
                }

                if (jiu[i / 3 * 3 + j / 3][num]) {
                    return false;
                }

                hang[i][num] = true;
                lie[j][num] = true;
                jiu[i / 3 * 3 + j / 3][num] = true;
            }
        }
        return true;
    }
};
// @lc code=end
